courses/fundamentals_of_ada/intro_160_genericity.rst

```diff
@@ -0,0 +1,394 @@
+   type T4 is private; -- indefinite
+   type T5 (<>) is private; -- indefinite
+   type T6 is tagged;
+   type T7 is array (Boolean) of Integer;
+   type T8 is access Integer;
+   type T9 is limited private;
+
+* Not limited to those choices
+
+.. code:: Ada
+
+   type T is not null access all limited tagged private;
+
+------
+Quiz
+------
+
+Which of the following statement is true?
+
+A. :Generics contracts define new types
+B. Generic contracts can express any type constraint
+C. :answer:`Generic contracts can express inheritance constraint`
+D. Generic contracts can require a type to be numeric (Real or Integer)
+
+.. container:: animate
+
+   A. No, the formal type and the actual type just have different views
+   B. Counter-example: representation clauses
+
+------
+Quiz
+------
+
+.. include:: quiz/generic_subp_syntax/quiz.rst
+
+------
+Quiz
+------
+
+.. code:: Ada
+
+   generic
+      type T1 is (<>);
+      type T2 (<>) is private;
+   procedure G
+     (A : T1;
+      B : T2);
+
+Which is an illegal instantiation?
+
+   A. :answermono:`procedure A is new G (String, Character);`
+   B. ``procedure B is new G (Character, Integer);``
+   C. ``procedure C is new G (Integer, Boolean);``
+   D. ``procedure D is new G (Boolean, String);``
+
+.. container:: animate
+
+   :ada:`T1` must be discrete - so an integer or an enumeration. :ada:`T2` can be any type
+
+=====================
+Generic Formal Data
+=====================
+
+----------
+Examples
+----------
+
+.. include:: examples/160_genericity/generic_formal_data.rst
+
+:url:`https://learn.adacore.com/training_examples/fundamentals_of_ada/160_genericity.html#generic-formal-data`
+
+--------------------------------------------
+Generic Constants and Variables Parameters
+--------------------------------------------
+
+.. container:: columns
+
+ .. container:: column
+
+    * Variables can be specified on the generic contract
+    * The mode specifies the way the variable can be used:
+
+       - :ada:`in` |rightarrow| read only
+       - :ada:`in out` |rightarrow| read write
+
+    * Generic variables can be defined after generic types
+
+ .. container:: column
+
+    .. code:: Ada
+
+       generic
+          type T is private;
+          X1 : Integer;  -- constant
+          X2 : in out T; -- variable
+       procedure P;
+
+       V : Float;
+
+       procedure P_I is new P
+          (T  => Float,
+           X1 => 42,
+           X2 => V);
+
+-------------------------------
+Generic Subprogram Parameters
+-------------------------------
+
+* Subprograms can be defined in the generic contract
+* Must be introduced by :ada:`with` to differ from the generic unit
+
+   .. code:: Ada
+
+      generic
+         with procedure Callback;
+      procedure P;
+      procedure P is
+      begin
+         Callback;
+      end P;
+      procedure Something;
+      procedure P_I is new P (Something);
+
+------
+Quiz
+------
+
+.. include:: quiz/genericity_type_and_variable/quiz.rst
```

Comment:
This code could be hard to read as-is (small text). If the idea is to just have it to "zoom in" and see, then it's fine, because the quality is preserved.

Reply:
I opened an issue for this kind of examples size issue in general https://github.com/AdaCore/training_material/issues/336